text/0025-configuration-policy.md

```diff
@@ -0,0 +1,197 @@
+      const queueProps = <QueueProps>props;
+      if (!id.includes("DLQ") && queueProps.deadLetterQueue === undefined) {
+        const dlq = new Queue(this, id + "DLQ");
+        return {
+          deadLetterQueue: { queue: dlq, maxReceiveCount: 3 },
+          ...props,
+        };
+      }
+    }
+    return props;
+  }
+}
+```
+
+# Motivation
+
+Teams, organizations, and companies may have need to enforce or apply default
+best practices to CDK-managed infrastructure. This approach will provide a
+consistent mechanism to apply these practices, while keeping the actual
+business-logic-containing Stacks lightweight.
+
+See the discussion in <https://github.com/aws/aws-cdk/issues/3235> for more
+context.
+
+# Basic Example
+
+_See README section._
+
+# Design Summary
+
+The approach adds a new hook (`onConstructCreation`) to the `Stack` class,
+which will be called by every L2 on construction. This hook receives the same
+inputs as the Construct's constructor, and enables users to adjust or override
+any of the input props, as well as throw errors or create ancillary resources.
+
+# Detailed Design
+
+A method is added to the `Stack` class with the following signature and default
+implementation:
+
+```ts
+public onConstructCreation(className: string, id: string, props: any): any {
+  return props;
+}
+```
+
+The `className` parameter will take the form of `module:class` (e.g.,
+`aws-s3.Bucket` or `aws-sqs.Queue`). The `id` and `props` parameters will be
+identical to the same parameters passed into the `Construct` on creation.
+
+All L2s will need to call this hook as part of their object construction. The
+call should be made immediately after the call to `super`, prior to property
+validation or Cfn* resource construction, as the modified properties may alter
+the outcome of these calls.
+
+Having the hook return the (unaltered or modified) properties is necessary
+because many of the FooProps interfaces contain readonly variables
+(e.g., `BucketProps.encryption`).
+
+This approach allows for injecting or enforcing default configuration, even in
+the cases of nested Constructs. For example, when creating a
+DnsValidatedCertificate, the properties of the underlying Lambda Function can
+be altered. However, no special consideration -- for example, an extra
+parameter -- is made to explicitly identify these Constructs as being nested.
+
+# Drawbacks
+
+The largest drawback to this change is the scope of the change -- every L2 will
+need to be extended to call the new hook as part of its initialization. Closely
+related to this, the pattern chosen does not lend itself well to the current
+static enforcement tools available (e.g., `awslint`) to ensure consistent
+behavior across the L2s. A possible mitigation might be to invest in a runtime
+linting capability across the CDK; however, this will expand the scope of the
+implementation considerably.
+
+Given the consistency in the change -- always the same one-liner after
```

Comment:
@rix0rrr has experimented with TS code injection... Maybe that's an opportunity to play with it a bit more.

I think that we must enforce this some how... Otherwise it's going to be too hard to get this right.

Reply:
Where would the code injection live? In jsii? 

Code injection is such a beautiful and dangerous approach. We could use it to actually add the call to the hook itself, or just enforce that a call is present. The former is a one-time effort that will scale as the CDK grows, but limits code discoverability; the advantage of the latter is that the call would have to still be present in source, but it does add the onus on each new construct author to add the call in by hand. Is there a precedence for this? Other areas of the CDK/jsii where we use injection to enforce standards and/or ensure call flow?

@rix0rrr - I would definitely be interested in your take on this.

Reply:
My take is thusly -- I am tired of linters that should have been fixers. We get slowed down too much with "put a comma here" and "put this text in the README", which are all things the computer should have been doing for me in the first place.

So I guess verify is good, fix it for me is better.